### ISSUE TITLE
False-negative type-checking Error and Invalid Suggestion

### ISSUE THREAD

**Issue by literally-user** [rust-lang/rust#162962]
I tried this code:

```rust
use std::ops::Neg;

struct Foo;

fn foo() -> Result<Foo, Box<dyn std::error::Error>> {
    Ok(Foo)
}

impl Neg for Foo {
    type Output = Result<Foo, Box<dyn std::error::Error>>;
    
    fn neg(self) -> Self::Output {
        Ok(self)
    }
}

fn bar() -> Result<Foo, Box<dyn std::error::Error>> {
    -foo()?
}

fn main() {
    bar().expect("bar should not fail");
}
```

I expected to see this happen: I receive `Result<Foo, E>` from `foo()` and `?` converts it to `Foo`, then i calling `neg()` from `Neg` trait and returning `Result<Foo, E>` from `bar()`

Instead, this happened: Type-check fails with `E0308` error and absolutely invalid suggestion in `help` section.

### Meta

`rustc --version --verbose`:
```
rustc 1.98.1 (48a229cea 2026-09-01)
binary: rustc
commit-hash: 48a229ceaefd4985c50990b14116b6d856af0985
commit-date: 2026-09-01
host: x86_64-unknown-linux-gnu
release: 1.98.1
LLVM version: 22.1.8
```

<details><summary>Backtrace</summary>
<p>

```
error[E0308]: `?` operator has incompatible types
  --> src/main.rs:18:6
   |
17 | fn bar() -> Result<Foo, Box<dyn std::error::Error>> {
   |             --------------------------------------- expected `Result<Foo, Box<(dyn std::error::Error + 'static)>>` because of return type
18 |     -foo()?
   |      ^^^^^^ expected `Result<Foo, Box<dyn Error>>`, found `Foo`
   |
   = note: `?` operator cannot convert from `Foo` to `Result<Foo, Box<(dyn std::error::Error + 'static)>>`
   = note: expected enum `Result<Foo, Box<(dyn std::error::Error + 'static)>>`
            found struct `Foo`
help: try removing this `?`
   |
18 -     -foo()?
18 +     -foo()
   |
help: try wrapping the expression in `Ok`
   |
18 |     -Ok(foo()?)
   |      +++      +

For more information about this error, try `rustc --explain E0308`.
error: could not compile `bug` (bin "bug") due to 1 previous error
```

</p>
</details>

**Comment by theemathas** [comment]
@lcnr This seems really wonky

**Comment by theemathas** [comment]
As a workaround, you can write
```rust
fn bar() -> Result<Foo, Box<dyn std::error::Error>> {
    let x = -foo()?;
    x
}
```

**Comment by mu001999** [comment]
I'd like to investigate this.

@rustbot claim
